feature_engine/timeseries/forecasting/window_features.py

```diff
@@ -0,0 +1,270 @@
+        pandas `shift()`.
+
+    sort_index: bool, default=True
+        Whether to order the index of the dataframe before creating the lag features.
+
+    {missing_values}
+
+    {drop_original}
+
+    Attributes
+    ----------
+    variables_:
+        The group of variables that will be used to create the window features.
+
+    {feature_names_in_}
+
+    {n_features_in_}
+
+    Methods
+    -------
+    {fit}
+
+    transform:
+        Add window features.
+
+    {fit_transform}
+
+    See Also
+    --------
+    pandas.rolling
+    pandas.aggregate
+    pandas.shift
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+        window: Union[str, int, Callable, List[int], List[str]] = 3,
+        min_periods: int = None,
+        functions: List[str] = ["mean"],
+        periods: int = 1,
+        freq: str = None,
+        sort_index: bool = True,
+        missing_values: str = "raise",
+        drop_original: bool = False,
+    ) -> None:
+
+        if isinstance(window, list) and len(window) != len(set(window)):
+            raise ValueError(f"There are duplicated windows in the list: {window}")
+
+        if not isinstance(functions, list) or not all(
+            isinstance(val, str) for val in functions
+        ):
+            raise ValueError(
+                f"functions must be a list of strings. Got {functions} instead."
+            )
+        if len(functions) != len(set(functions)):
+            raise ValueError(
+                f"There are some duplicated functions in the list of "
+                f"functions: {functions}"
+            )
+
+        if not isinstance(periods, int) or periods <= 0:
+            raise ValueError(
+                f"periods must be a positive integer. Got {periods} instead."
+            )
+
+        super().__init__(variables, missing_values, drop_original)
+
+        self.window = window
+        self.min_periods = min_periods
+        self.functions = functions
+        self.periods = periods
+        self.freq = freq
+        self.sort_index = sort_index
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Adds window features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: Pandas dataframe, shape = [n_samples, n_features + window_features]
+            The dataframe with the original plus the new variables.
+        """
+        # Common dataframe checks and setting up.
+        X = super().transform(X)
+
+        if isinstance(self.window, list):
+            df_ls = []
+            for win in self.window:
+                tmp = (
+                    X[self.variables_]
+                    .rolling(window=win)
+                    .agg(self.functions)
+                    .shift(periods=self.periods, freq=self.freq)
```

Comment:
What is the use case to shift the rolling window feature by greater than one period? If there isn't one, then removing it could help simplify this class.

Reply:
I think it is a fair point.

I just worry that some users will be using pandas `periods` mostly for shift, and then obliging them to use `freq` instead must be disorienting/frustrating.

Also, supposedly, pandas also shifts and rolls even if we do not have a datetime in the index. So in those cases, freq won't be useful.